openssl/src/pkey.rs

```diff
@@ -216,18 +234,28 @@ impl PKeyCtxRef {
         };
         Ok(Padding::from_raw(pad))
     }
-}
 
-impl ForeignTypeRef for PKeyCtxRef {
-    type CType = ffi::EVP_PKEY_CTX;
+    pub fn derive_from_peer(&mut self, peer: &PKeyRef) -> Result<Vec<u8>, ErrorStack> {
+        unsafe {
+            try!(cvt(ffi::EVP_PKEY_derive_init(self.as_ptr())));
```

Comment:
Seems like it might make sense to define a method for each of these functions? From reading the man pages, it seems like there are some things you may want to do after setting the peer at the very least.